### ISSUE TITLE
Discussion: Possible solutions for the BAG onderzoeken relations

### ISSUE THREAD

**Issue by JKoetsier** [Amsterdam/GOB-Upload#695]
**The challenge**
In BAG we have a new onderzoeken collection. This collection has an attribute 'object_identificatie' that points to one of 7 other BAG object classes: the object_identificatie for a an onderzoeken object can point to a ligplaats, standplaats, woonplaats, or ....

_The (current) GOB-way_
In GOB, the usual way to define a relation is to define a source value ('bronwaarde') on import. For a single Reference this means that, for an object A that has a relation to B, one source value is defined. For a ManyReference, multiple source values can be defined. However, for every single source value we only expect one and only one relation to B. If either no relations are found for a given source value, or, on the other hand, more than 1 relation are found for a given source value, GOB will generate a data warning.
Also, for every relation in GOB we define the destination object class. For example, if we have an object A with a relation r1 to object class B, we define an attribute R1 that holds the source value(s) for this relation. If, however, we also want A to have a relation with object class C, we would need to define a new attribute R2 that holds the source values for relation r2. Even if the source values used for the relation Ar1B and Ar2C are the same (R1 = R2), we still need to define to separate attributes R1 and R2, because we can only define one relation per attribute.

_Relating onderzoeken the GOB-way_
Currently, there are 2 options for implementing onderzoeken in GOB, both are less than ideal:

- We can define 7 different attributes for all possible relations on the onderzoeken object. This means that for all objects in onderzoeken, all 7 attributes contain the same bronwaarde (the object_identificatie), but only 1 of all 7 relations will be found. GOB will log data warnings for the other 6 relations. We do something similar in stukdelen. Not a nice solution. 

- We define a relation to onderzoeken on all 7 BAG object classes. This is much cleaner than the other option, but we need to know upfront how many onderzoeken relations we expect. As mentioned earlier, we need to define as many source values as expected relations. This means we need to mimick the relate process during the import, to prefill the relation attributes with the expected onderzoeken. We tried this approach earlier, but this was suboptimal for several reasons; missing indexes on the source database, getting the relate logic right in the import, etc. It will definitely work, but we're looking for a more durable solution. A small downside to this solution too is that if onderzoeken is updated, the relations may lag behind, as all objects need to be updated too. It's not possible to only update onderzoeken and run relate with this solution.

**Possible solutions**

_1. Accepting multiple relations for one source value_
We could define the onderzoeken relation on the 7 objects and ignore the 'one source value = one relation' rule with a flag in gobmodel. We use '=identificatie' as bronwaarde (just as we do with '=geometrie', which says "take the value of my attribute") and relate should work as it does now. There is, as far as I can see now, only one thing we have to solve with this approach, and that is that we need to generate the relation ID's for these kinds of relations a bit differently. Currently a relation in the relation table is identified with 'src_id.src_volgnummer.bronwaarde'. If multiple matches are possible on a single bronwaarde, we would have to add the dst_id.dst_volgnummer to this as well. This makes the relation unique in the relation table, and keeps everything else working as it should.
This change is limited to a very small part of relate and check relate. It's basically just assigning different ID's to this kind of relations and returning all matches instead of only the first.

_2. The ability to define multiple possible destination collections_
We would only have to define the relation on onderzoeken, with the 7 possible destination collections. I think however that the impact of such a change would be massive, as GOB is littered with logic that acts on a dst catalog and collection. I don't think this is really an option.

_3. Reversed relations_
A solution Jurien proposed. I'll let Jurien explain this one, as it's not yet clear to me ;)

_4. ..._

_5. ..._

**Comment by Vegter** [comment]
**Relations in GOB**

Relations in GOB are of 3 kinds: Reference, ManyReference, VeryManyReference. The type of the relation stands for the cardinality (An alternative (better) approach would probably be to have one type: Reference and the cardinality as a property).

Suppose:
- A has a reference to B.
- Both A and B are uniquely identified by their id (identification + sequence number).
- A has some property x (bronwaarde) that tells how it refers to B.

The relate process will resolve the property x to the id of the B that A refers to.
The outcome is a relation and relation between A and B that is valid during a time interval:
- {A.id, A.bronwaarde => A, B, from, to}

The relate process also checks the integrity of the relations:
- If A does not specify the bronwaarde (x) the relation is reported as unresolvable: _“source value should exist”_
- If no B can be found for A then the relation is reported as dangling: _“referenced entity should exist”_
- If more than one B is found for an A reference the relation is reported as ambiguous: _“destination is ambiguous”_

The relate check clearly states the problem:
If A refers to B then A has a property x (a code, its geometry, …) that is resolved by comparing the property x to a property of B. If a match is found the relation is established.
If the bronwaarde is missing, if no B is found, or if more than one B is found for the same bronwaarde the relation is marked as erroneous.
The whole relation algorithm is build upon the fact that A has bronwaardes, and that every bronwaarde corresponds to one B entity; A points to B. The bronwaarde of A tells something about B.

The question is: can this process be reversed?
The bronwaarde of A tells something about something that B knows about A.
In fact relate no longer resolves A => B but B => A.

In fact this is pretty simple. One could even argue that relate already does this but that is not entirely true.

**One approach:**
Just solve the reverse relation B => A. However, not all B’s will relate to A and the relate check might  issue “referenced entity should exist” errors. Also, the relations should be registered with the unique identification of the B – A relation, but as a A - B relation.

A nice approach, simple in it’s idea but the wording is probably already confusing so probably we should leave this approach aside for now.

**Other approach:**
Use the regular relate process but identify reverse relations. This identification of reverse relations is required because of the following reasons.

First of all the following relate checks will fail:
It is not guaranteed that B’s exist that refer to A and this will trigger the _“referenced entity should exist”_ error.
Also, more B’s can refer to the same A and this will trigger the _“destination is ambiguous”_ error.

Meta information about the reference is required to suppress these checks. A proposed solution is to extend the resolve info in gobsourses with an extra attribute ‘**reversed**’ that tells the relate process that this relation is not an ordinary relation where A refers to B but a relation where all B’s should be found that refer to A.

The meta attribute is required to skip the checks only for reverse relations. The checks for the “normal” relations should remain unchanged. This solution will solve the relate check errors.

Another problem is the identification of the relation:
- {A.id, A.bronwaarde => A, B, from, to}

As there are many B’s that can refer to A using the same bronwaarde this will become:
- {A.id, A.bronwaarde => A, B1, from, to}
- {A.id, A.bronwaarde => A, B2, from, to}

The same bronwaarde from A defines 0 to many relations with B.

The relation is no longer uniquely identified so the unique identification has to be updated.

**One solution is to include the identification of B:**
The unique identification will then become:
- {A.id, A.bronwaarde, B.id => A, B, from, to}

It’s tempting to just extend the unique identification.
But if you look at it more closer it is just a reverse relation:
The key could also be changed to:
- {B.id, A.bronwaarde => A, B, from, to}

You might even include the word “reverse” in the key to avoid confusion with any “normal” B to A relation.

The seconde approach is more readable. If A tells how B refers to A (that is a reverse relation) then this will probably be by means of its id. The unique identification of the relation so becomes: “A.id, A.id, B.id” in the first approach and “B.id, A.id” in the second approach. The latter is easier to understand.

However, that is just a personal preference. Whatever solution is chosen: the unique identification has to guarantee that there exists maximally one B=>A relation for any combination A-B; A can have zero to many B’s that point to it but any B points to maximally one A.

**The new identification should only be used for “reversed” relations.**
The new unique identification could also be used as identification for all “normal” relations but would cause many change and delete events. Even more, a database constraint exists on the id that protects the integrity of the relations, so for “normal” relations the id should stay as it is.

As a last step the relate logic has to be adjusted so that id’s for reverse relations are correctly calculated, to return multiple results for reverse relations (remove the LIMIT 1 clause) and to correctly process the results.

**Summary:**
There is a functional and technical difference between a relation where:
- A knows something about  an attribute of B that relates A to B
or
- A knows something about an attribute of B that relates B to A

The outcome is however identical: a set of A-B relations.

The relate process has to be updated to handle this new type of relationship.
Basically this comes down to:
- Add meta information to the relation specs in gobsources to identify the new relation type
- Use this info to adjust the relate query and relate checks
- Use a new identification key for reversed relations

**Comment by Robkromwijk** [comment]
As stated the outcome is a set of A-B relations. Also the informationmodel is the same: A has a reference to B. Even cardinality and optionality could be added.

**Model**
```
"heeft_onderzoeken": {
   "type": "GOB.Reference",
   "cardinality: "many",
   "optional": "true",
   "description": "object kan in onderzoek zijn",
   "ref": "bag.onderzoeken"
}
```

By adding the direction of the relation when determining, it should be possible to get the set of A-B relations by performing a B-A lookup.

**Config**
```
"heeft_onderzoeken": {
   "source_mapping": {
      "bronwaarde": "object_identificatie",
      "direction": "reverse"
}
```
